internal/provider/newrelic/provider_integration_test.go

```diff
@@ -168,12 +181,228 @@ func Test_Getting_external_metric_through_proxy(t *testing.T) {
+					w.WriteHeader(http.StatusOK)
+				},
+				wantErr: true,
+			},
+			"is_400": {
+				handlerFunc: func(w http.ResponseWriter, _ *http.Request) {
+					w.WriteHeader(http.StatusBadRequest)
+				},
+				wantErr: true,
+			},
+			"is_401": {
+				handlerFunc: func(w http.ResponseWriter, _ *http.Request) {
+					w.WriteHeader(http.StatusUnauthorized)
+				},
+				wantErr: true,
+			},
+			"is_403": {
+				handlerFunc: func(w http.ResponseWriter, _ *http.Request) {
+					w.WriteHeader(http.StatusForbidden)
+				},
+				wantErr: true,
+			},
+			"is_500": {
+				handlerFunc: func(w http.ResponseWriter, _ *http.Request) {
+					w.WriteHeader(http.StatusInternalServerError)
+				},
+				wantErr: true,
+			},
+			"is_200_for_first_requests_and_then_500": {
+				handlerFunc: func(w http.ResponseWriter, r *http.Request) {
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusOK)
+					_, _ = w.Write([]byte(`{"data": {
+				"actor": {
+				  "account": {
+					"nrql": {
+					  "results": [{"one": 0.015}]
+					}
+				  }
+				}
+			}}`))
+				},
+				wantErr:           false,
+				makeExtraRequests: true,
+			},
+		}
+
+		for testCaseName, testData := range cases {
+			testData := testData
+			//nolint:paralleltest
+			t.Run(testCaseName, func(t *testing.T) {
+				server := runHTTPTestServer(t, 3000, testData.handlerFunc)
+
+				m := provider.ExternalMetricInfo{Metric: testMetricName}
+
+				_, err := p.GetExternalMetric(ctx, "", nil, m)
+
+				if testData.wantErr {
+					require.Error(t, err)
+					testutil.RetryGetRequestAndCheckStatus(ctx, t, testEnv.HTTPClient, url, func(statusCode int) bool {
+						return statusCode == http.StatusOK
+					})
+				} else {
+					require.NoError(t, err)
+					testutil.RetryGetRequestAndCheckStatus(ctx, t, testEnv.HTTPClient, url, func(statusCode int) bool {
+						return statusCode != http.StatusOK
+					})
+				}
+
+				if testData.makeExtraRequests {
+					// Make second request.
+					if _, err := p.GetExternalMetric(ctx, "", nil, m); err != nil {
+						t.Errorf("Unexpected error = %v", err)
+					}
+
+					testutil.CheckStatusCodeOK(ctx, t, testEnv.HTTPClient, url)
+
+					// Make the third request to fail.
+					server.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+						w.Header().Set("Content-Type", "application/json")
+						w.WriteHeader(http.StatusInternalServerError)
+					})
+
+					if _, err := p.GetExternalMetric(ctx, "", nil, m); err == nil {
+						t.Error("Error expected")
+					}
+
+					checkStatusCodeNotOK(ctx, t, testEnv.HTTPClient, url)
+				}
+			})
+		}
+	})
+}
+
+// checkStatusCodeNotOK sends a GET request to the given URL.
+//
+// If 401 or 403 return code is received, function will retry.
+//
+// If a status code = 2xx is received, function will fail the given test.
+func checkStatusCodeNotOK(ctx context.Context, t *testing.T, httpClient *http.Client, url string) {
```

Comment:
I think that disabling the cache would make this function much more simpler.

Then a separate test for when the cache is enabled could be added but just doing some time.sleep(for a given time > cache life) instead of retrying.

Reply:
Answered here https://github.com/newrelic/newrelic-k8s-metrics-adapter/pull/52#discussion_r849361451.